### ISSUE TITLE
Universal CLI tool for all

### ISSUE THREAD

**Issue by LFlare** [zerotier/ZeroTierOne#521]
Hello,

I was wondering, is there any development under way for a CLI tool that interfaces with the entire JSON API including the network controller? I was thinking of developing one myself for Python...

EDIT: I've tried making my own [tool](https://github.com/LFlare/zerotiernc), though I doubt it's 100% good...

**Comment by soakes** [comment]
Hi There,

I am also looking for the same which was ideally pre-made already and the only ones I can find are in dead links from a post from https://mangolassi.it/topic/5851/zerotier-review/21 by a guy called "dafyre" on "1 Sep 2015, 13:34"

Does anyone have a copy of these files? or some python, bash, php CLI management tools? that I would take a peek at?

TIA.

**Comment by LFlare** [comment]
@soakes, the network JSON API itself is fairly simple and I've replicated most of it in my Python script but I was wondering if there were any **official** implementation...

**Comment by soakes** [comment]
@LFlare thank you so much for providing the python script, it likely will do my needs, so a big thank you again.  Just about to download it and take a peak.

**Comment by dafyre** [comment]
Nice writeup, @LFlare!

**Comment by LFlare** [comment]
@soakes Apart from network rules, I do believe I covered everything. Please let me know if you need help using it, as the guide hasn't been written. 

@dafyre What writeup?

**Comment by dafyre** [comment]
Sorry, comment fixed.

**Comment by soakes** [comment]
@LFlare will do, but should be simple enough as the python modules are added at the top of the file.  but will yell if I get absolutely stuck, thank you.

**Comment by LFlare** [comment]
@soakes It's mostly the argument and command implementation that I'm worried about.

**Comment by soakes** [comment]
@LFlare yeah there is no help parameters but I do understand a little python so I should be able to muddle though.  

All I really need to do is:

- create new network
- assign manual IP range to be used
- add one managed route for that CIDR range as the rest are fead via OSPF.
- auth new networks

I don't know how the relay part works so I need to read up on that so theres likely some bits ive missed.  This is just basically going by the my.zerotier.com site.

**Comment by LFlare** [comment]
@soakes There is a default `-h` option that will display some help information. In any case, for your case,

- `./zerotier-nc.py --net-add -n {optional 6-char suffix}`
- `./zerotier-nc.py --net-ipadd {cidr} -n {network id}`
- `./zerotier-nc.py --member-list -n {network id}`
- `./zerotier-nc.py --member-auth -n {network id} -z {member id}`

There are also useful functions I added like aliasing networks and members into the form of `networkalias:memberalias`.

- `./zerotier-nc.py --alias lan -n {network id}`
- `./zerotier-nc.py --alias macbook -z {member id} lan`
- `./zerotier-nc.py --member-info lan:macbook`

Just a little something I made because having to remember IDs is a pain.

**Comment by soakes** [comment]
Thats perfect, thank you.

I must of missed it in the code, but do admit I only briefly flicked though it as I am not quite ready to test it all yet as the controller hasn't been built yet plus i'm semi busy with other work stuff right now.  I was planning to go though it fully when the controller was/is built.  

Right now I am just finishing of the rest of my Moons and then will start work on developing the new network (controller etc) which is when I need your awesome script.

**Comment by soakes** [comment]
@LFlare I think your right, recently its been moved so its installed by default now.  I know the docs I was reading was telling you to recompile it with a certain flag, this I believe is no longer needed.

In either case, I need to create a dev network which isn't on the production one for testing to make sure own controller works as it should (thats more to do with configuring it).

**Comment by LFlare** [comment]
@soakes Yeah, ZeroTier is seriously cool but it also appears that there's only 1-2 guys behind the actual project so documentations and manuals are all slightly outdated. A slight tip though, never attempt to `kill -9` or `SIGKILL` the binary when it's running as a controller as well, seems like there's a bug that causes the entire installation to die. Do a termination by default `kill` or `SIGTERM`.

**Comment by soakes** [comment]
@LFlare Thanks very much for that TIP.

**Comment by soakes** [comment]
@LFlare Just a quick question while I think about it as i'm in the middle of planning the new network.   I have about 50-60 networks (theres more devices behind these, so that number is far from the actual total devices) but this is all thats directly connected if you will.  

Now I am trying to plan what would be the best way to go ahead with creating controllers, roots etc.  So my question to you is, how many controllers and moons do you suggest?  I understand that the roots (moons) need to be on separate boxes, but should they be on the same ISP as the controllers or fairly near on another ISP?

I have mostly three major DC locations, UK, DE, SK and misc VPS dotted around the world via different ISPs but they are mostly in/around Europe.  So with this in mind, how many controllers, moons and whatever else you think else that I need, what would you suggest? 

Or is it just one controller + many moons? bit lost :)

**Comment by LFlare** [comment]
@soakes If you read the [manual](https://www.zerotier.com/manual.shtml), it recommends that the moons have fairly good network connections, which might also mean that the closer the better, as they function as secondary root servers and initial latency in looking up network members might play a role.

Conversely, controllers by themselves don't take part in the whole network themselves, they are just middlemans responsible for keeping records of members, so you won't have to worry about the number of controllers online or geographical location.

So, to sum it up, you will need a minimum of **2 moons** (or 1 if you are feeling frisky) and a minimum of **1 controller**. Honestly, there's no need to make multiple moons or controllers as the implementation of the protocol itself is performance-streamlined. 

Fun Fact: ZeroTier Central by itself uses **7** network controllers but that's a given if you take into account the number of networks they have to handle.

Extra Fun Fact: ZeroTier root servers number **12** in total as of 2015, may be more now but I doubt it.

TLDR: Unless you plan to make OneTier or another mock-up, you shouldn't need more than 1/2 moons and 1 controller.

**Comment by soakes** [comment]
@LFlare Thats brilliant thank you for explaining that.  I did read the bit about latency/location which is why I was asking but nothing really said about ratios etc.

This makes it much easier to implement, thank you.

Moons are covered and already made, now just the controller configuration/machine.

**Comment by LFlare** [comment]
@soakes Also, join the #zerotier channel on Freenode for live support and chat =3

**Comment by soakes** [comment]
Awesome will do.   Busy day at work so no chance today/
